packages/workspace/src/browser/workspace-frontend-contribution.ts

```diff
@@ -106,47 +138,135 @@ export class WorkspaceFrontendContribution implements CommandContribution, Keybi
-            title: WorkspaceCommands.OPEN_WORKSPACE.label!,
-            canSelectFiles: false,
+    /**
+     * Opens a folder after prompting the `Open Folder` dialog. Resolves to `undefined`, if
+     *  - the workspace root is not set,
+     *  - the folder to open does not exist, or
+     *  - it was not a directory, but a file resource.
+     *
+     * Otherwise, resolves to the URI of the folder.
+     */
+    protected async doOpenFolder(): Promise<URI | undefined> {
+        const props: OpenFileDialogProps = {
+            title: WorkspaceCommands.OPEN_FOLDER.dialogLabel,
             canSelectFolders: true,
+            canSelectFiles: false
         };
-        await this.preferences.ready;
-        if (this.preferences['workspace.supportMultiRootWorkspace']) {
-            option.canSelectFiles = true;
-            option.filters = {
-                'Theia Workspace (*.theia-workspace)': [THEIA_EXT],
-                'VS Code Workspace (*.code-workspace)': [VSCODE_EXT]
-            };
+        const [rootStat] = await this.workspaceService.roots;
+        const destinationFolderUri = await this.fileDialogService.showOpenDialog(props, rootStat);
+        if (destinationFolderUri) {
+            const destinationFolder = await this.fileSystem.getFileStat(destinationFolderUri.toString());
+            if (destinationFolder && destinationFolder.isDirectory) {
+                this.workspaceService.open(destinationFolderUri);
+                return destinationFolderUri;
+            }
         }
-        const selected = await this.fileDialogService.showOpenDialog(option);
-        if (selected) {
-            // open the selected directory, or recreate a workspace from the selected file
-            this.workspaceService.open(selected.uri);
+        return undefined;
+    }
+
+    /**
+     * Opens a workspace after raising the `Open Workspace` dialog. Resolves to the URI of the recently opened workspace,
+     * if it was successful. Otherwise, resolves to `undefined`.
+     *
+     * **Caveat**: this behaves differently on different platforms, the `workspace.supportMultiRootWorkspace` preference value **does** matter,
+     * and `electron`/`browser` version has impact too. See [here](https://github.com/theia-ide/theia/pull/3202#issuecomment-430884195) for more details.
+     *
+     * Legend:
+     *  - `workspace.supportMultiRootWorkspace` is `false`: => `N`
+     *  - `workspace.supportMultiRootWorkspace` is `true`: => `Y`
+     *  - Folders only: => `F`
+     *  - Workspace files only: => `W`
+     *  - Folders and workspace files: => `FW`
+     *
+     * -----
+     *
+     * |---------|-----------|-----------|------------|------------|
+     * |         | browser Y | browser N | electron Y | electron N |
+     * |---------|-----------|-----------|------------|------------|
+     * | Linux   |     FW    |     F     |     W      |     F      |
+     * | Windows |     FW    |     F     |     W      |     F      |
```

Comment:
In the table above, Electron with multi-root = Y, it should be "FW" and not jut "W" for Linux environment
Cannoit test the other environment (Window and OSX)

Reply:
> it should be "FW"

What do you mean by it "should be"? Did you mean you can open both the workspace files and folders on Linux in electron?